### PR TITLE
Add variable definition for enum types

### PR DESCRIPTION
**Pull request by Dakror**
I got tired of having to register custom getter and setters for every enum type i have in my structs, so i made this small utility change using std::underlying_type

### CODE REVIEW

File: Include/RmlUi/Core/DataVariable.h [hunk 1]
```diff
@@ -111,6 +111,24 @@ class ScalarDefinition final : public VariableDefinition {
 };
 
 
+template<typename T>
+class EnumDefinition final : public VariableDefinition {
```
Comment (mikke89) [inline Include/RmlUi/Core/DataVariable.h:115]: Isn't this equivalent to `ScalarDefinition<std::underlying_type<T>::type>`, can we just re-use that type?

File: Include/RmlUi/Core/DataTypeRegister.h [hunk 2]
```diff
@@ -94,10 +94,10 @@ class RMLUICORE_API DataTypeRegister final : NonCopyMoveable {
 
 		return definition.get();
 	}
-
-	// Get definition for types that are not a built-in scalar.
-	// These must already have been registered by the user.
-	template<typename T, typename std::enable_if<!PointerTraits<T>::is_pointer::value && !is_builtin_data_scalar<T>::value, int>::type = 0>
+	
+	       // Get definition for types that are not a built-in scalar.
```
Comment (mikke89) [inline Include/RmlUi/Core/DataTypeRegister.h:98]: Some weird formatting going on here?

File: Include/RmlUi/Core/DataTypeRegister.h [hunk 3]
```diff
@@ -94,10 +94,10 @@ class RMLUICORE_API DataTypeRegister final : NonCopyMoveable {
...
+	       // These must already have been registered by the user.
+	template<typename T, typename std::enable_if<!PointerTraits<T>::is_pointer::value && !is_builtin_data_scalar<T>::value && !std::is_enum<T>::value, int>::type = 0>
```
Comment (mikke89) [inline Include/RmlUi/Core/DataTypeRegister.h:100]: I think it would make sense to make it part of our "built-in data scalar" concept.

**Review by mikke89 — COMMENTED** [review]
Thanks, this looks useful to me!

**Comment by mikke89** [comment]
I like this approach. A getter would be nice too, so we could use it for retrieving property values. Currently we set all enums using int as the type on properties, not sure if this could cause any issues.

I hope the library-wide formatting I decided to do isn't causing you too much trouble here. At least now you should be able to enable "format on save" if you have clang-format enabled on your editor.

There is also a large file here I don't think you intended to add.

**Comment by Dakror** [comment]
oops yeah the compilation database..

**Comment by Dakror** [comment]
I actually dont know what the best way for a setter would be, i dont want to be too invasive in the variant type. do you have an idea for it?

**Comment by mikke89** [comment]
I think your solution there already is good. Maybe, to be more generic, cast it to a specific integer type (int64_t?) before storing. Otherwise, if we don't support a specific underlying type, I suspect it won't compile. static_assert the size to be sure it fits in the int64_t. I think this should be safe?

Did you mean the getter? I think the same principle, we would need an SFINAE for GetInto enums. Maybe it is a good idea to measure compile times before and after this change :)

**Comment by Dakror** [comment]
image
im kinda stuck, dont know what im doing wrong in my getter... ideas?

**Comment by Dakror** [comment]
I worked around it using reinterpret_cast, not sure if thats a good idea though.

**Comment by Dakror** [comment]
True thats a good point. Thanks for sharing your solution, i kinda figured using a temporary value for the GetInto function would be a version that could work.

Think we can just use your version then

**Comment by mikke89** [comment]
Alright, sounds good. Do you want to integrate that solution into your other changes here?

**Comment by mikke89** [comment]
I wanted a few more changes and just went ahead and committed them, hope that's okay. If you are happy with this then I am ready to merge it.

**Comment by Dakror** [comment]
looks good to me

**Comment by mikke89** [comment]
Alright, thanks again for the PR.

**Pull request merged**